STITCH-1366 - Implement Twilio Service in Core, Android, and Server

edaniels opened the pull request:
In addition to Twilio...

- Clean up some tests
- Add some more tests
- Refactor packages to support publishing
- Allow direct construction of credentials to streamline auth flow
- Rename ClientProvider's to ClientFactory's
- Add mongodb-local to server
- Change Blocking suffix to Internal to reflect what it's really for
- Make builders more fluent (at the cost of duplication)

edaniels commented:
@adamchel @jsflax please take a look when you get a chance and take your time. Happy to go over all of this in person too. It's a lot of changes but they are all pretty consistent and nothing should be too alarming.

edaniels commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchAuthDocRequest.java:
```diff
@@ -0,0 +1,165 @@
+    /**
+     * Set if this request should use an access token in this request.
+     */
+    public Builder withAccessToken() {
```
You'll see this pattern of overriding and using super (without the annotation) to indicate that this is purely for allowing you to do `builder.withSuperMethod().withConcreteMethod().withSuperMethod().build()` and still get back the concrete built type. CRTP (https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) didn't work out here and it confuses me very much in general :)

edaniels commented on android/services/twilio/src/androidTest/java/com/mongodb/stitch/android/services/twilio/TwilioServiceClientIntTests.kt:
```diff
@@ -0,0 +1,100 @@
+import org.junit.runner.RunWith
+import java.util.concurrent.ExecutionException
+
+@RunWith(AndroidJUnit4::class)
```
@jsflax here are the int tests

edaniels commented on core/services/twilio/src/test/java/com/mongodb/stitch/core/services/twilio/CoreTwilioServiceClientUnitTests.java:
```diff
@@ -0,0 +1,99 @@
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+
+public class CoreTwilioServiceClientUnitTests {
```
@jsflax and here are the unit tests.

jsflax replied:
👍

edaniels commented:
If we agree on changes here, we should make the changes on swift and js to maintain symmetry.

jsflax commented on android/services/twilio/src/androidTest/java/com/mongodb/stitch/android/services/twilio/TwilioServiceClientIntTests.kt:
```diff
@@ -0,0 +1,100 @@
+        val client = getAppClient(app.first)
+        Tasks.await(client.auth.loginWithCredential(AnonymousCredential()))
+
+        val twilio = client.getServiceClient(TwilioServiceClient.Factory, "twilio1")
```
Should be fine on Swift side as well

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/Response.java:
```diff
@@ -20,13 +20,15 @@
  */
 public final class Response {
-  private final Integer statusCode;
+  private final int statusCode;
```
Nice, always preferably to use primitives where possible, especially on mobile.

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchAuthRequest.java:
```diff
@@ -125,11 +127,67 @@ public Builder withShouldRefreshOnFailure(final boolean shouldRefresh) {
      * Builds the request.
      */
     public StitchAuthRequest build() {
+      if (useRefreshToken) {
```
[nice]

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchRequest.java:
```diff
@@ -190,4 +191,40 @@ public StitchRequest build() {
     }
   }
+
+  protected Document toDocument() {
```
[q] Why add this?

edaniels replied:
For tests. It makes for nice assertEquals :)

jsflax commented on core/services/twilio/src/main/java/com/mongodb/stitch/core/services/twilio/CoreTwilioServiceClient.java:
```diff
@@ -0,0 +1,59 @@
+    this.service = service;
+  }
+
+  private void sendMessageBlocking(
```
[nit][q] Why the blocking and internal?

edaniels replied:
whoops. It should be merged with the caller

jsflax commented on core/services/twilio/src/main/java/com/mongodb/stitch/core/services/twilio/TwilioService.java:
```diff
@@ -0,0 +1,23 @@
+
+package com.mongodb.stitch.core.services.twilio;
+
+public final class TwilioService {
```
[q] Where is this used?

edaniels replied:
I was thinking the admin SDK, but maybe this is better removed!

edaniels commented:
Addressed comments. I think the remote MDB service within core will actually work really well. It will let us define everything in blocking behavior and have all common logic there. The android side will really just be one large facade that implements an android interface by extending core and wrapping its calls in dispatchers.

edaniels commented:
@adamchel RFAL

edaniels commented on android/services/twilio/src/main/java/com/mongodb/stitch/android/services/twilio/TwilioServiceClient.java:
```diff
@@ -0,0 +1,69 @@
+   * @param body The body text of the message.
+   * @return A task that completes when the send is done.
+   */
+  Task<Void> sendMessage(
```
I agree, let's add those kinds later.
